diff --git a/include/v8-platform.h b/include/v8-platform.h
index 788122955fd..2decd2088b5 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -394,20 +394,22 @@ class PageAllocator {
     kNoAccess,
     kRead,
     kReadWrite,
     kReadWriteExecute,
     kReadExecute,
     // Set this when reserving memory that will later require kReadWriteExecute
     // permissions. The resulting behavior is platform-specific, currently
     // this is used to set the MAP_JIT flag on Apple Silicon.
     // TODO(jkummerow): Remove this when Wasm has a platform-independent
     // w^x implementation.
+    // TODO(saelo): Remove this once all JIT pages are allocated through the
+    // VirtualAddressSpace API.
     kNoAccessWillJitLater
   };
 
   /**
    * Allocates memory in range with the given alignment and permission.
    */
   virtual void* AllocatePages(void* address, size_t length, size_t alignment,
                               Permission permissions) = 0;
 
   /**
@@ -503,22 +505,73 @@ class PageAllocator {
    * INTERNAL ONLY: This interface has not been stabilised and may change
    * without notice from one release to another without being deprecated first.
    *
    * If not overridden and changed to return true, V8 will not attempt to call
    * AllocateSharedPages or RemapSharedPages. If overridden, AllocateSharedPages
    * and RemapSharedPages must also be overridden.
    */
   virtual bool CanAllocateSharedPages() { return false; }
 };
 
+// Opaque type representing a handle to a shared memory region.
+using PlatformSharedMemoryHandle = intptr_t;
+static constexpr PlatformSharedMemoryHandle kInvalidSharedMemoryHandle = -1;
+
+// Conversion routines from the platform-dependent shared memory identifiers
+// into the opaque PlatformSharedMemoryHandle type. These use the underlying
+// types (e.g. unsigned int) instead of the typedef'd ones (e.g. mach_port_t)
+// to avoid pulling in large OS header files into this header file. Instead,
+// the users of these routines are expected to include the respecitve OS
+// headers in addition to this one.
+#if defined(V8_OS_MACOSX) && !defined(V8_OS_IOS)
+// Convert between a shared memory handle and a mach_port_t referencing a memory
+// entry object.
+inline PlatformSharedMemoryHandle SharedMemoryHandleFromMachMemoryEntry(
+    unsigned int port) {
+  return static_cast<PlatformSharedMemoryHandle>(port);
+}
+inline unsigned int MachMemoryEntryFromSharedMemoryHandle(
+    PlatformSharedMemoryHandle handle) {
+  return static_cast<unsigned int>(handle);
+}
+#elif defined(V8_OS_FUCHSIA)
+// Convert between a shared memory handle and a zx_handle_t to a VMO.
+inline PlatformSharedMemoryHandle SharedMemoryHandleFromVMO(uint32_t handle) {
+  return static_cast<PlatformSharedMemoryHandle>(handle);
+}
+inline uint32_t VMOFromSharedMemoryHandle(PlatformSharedMemoryHandle handle) {
+  return static_cast<uint32_t>(handle);
+}
+#elif defined(V8_OS_WIN)
+// Convert between a shared memory handle and a Windows HANDLE to a file mapping
+// object.
+inline PlatformSharedMemoryHandle SharedMemoryHandleFromFileMapping(
+    void* handle) {
+  return reinterpret_cast<PlatformSharedMemoryHandle>(handle);
+}
+inline void* FileMappingFromSharedMemoryHandle(
+    PlatformSharedMemoryHandle handle) {
+  return reinterpret_cast<void*>(handle);
+}
+#else
+// Convert between a shared memory handle and a file descriptor.
+inline PlatformSharedMemoryHandle SharedMemoryHandleFromFileDescriptor(int fd) {
+  return static_cast<PlatformSharedMemoryHandle>(fd);
+}
+inline int FileDescriptorFromSharedMemoryHandle(
+    PlatformSharedMemoryHandle handle) {
+  return static_cast<int>(handle);
+}
+#endif
+
 /**
- * Page permissions.
+ * Possible permissions for memory pages.
  */
 enum class PagePermissions {
   kNoAccess,
   kRead,
   kReadWrite,
   kReadWriteExecute,
   kReadExecute,
 };
 
 /**
@@ -527,25 +580,27 @@ enum class PagePermissions {
  * This class represents a contiguous region of virtual address space in which
  * sub-spaces and (private or shared) memory pages can be allocated, freed, and
  * modified. This interface is meant to eventually replace the PageAllocator
  * interface, and can be used as an alternative in the meantime.
  */
 class VirtualAddressSpace {
  public:
   using Address = uintptr_t;
 
   VirtualAddressSpace(size_t page_size, size_t allocation_granularity,
-                      Address base, size_t size)
+                      Address base, size_t size,
+                      PagePermissions max_page_permissions)
       : page_size_(page_size),
         allocation_granularity_(allocation_granularity),
         base_(base),
-        size_(size) {}
+        size_(size),
+        max_page_permissions_(max_page_permissions) {}
 
   virtual ~VirtualAddressSpace() = default;
 
   /**
    * The page size used inside this space. Guaranteed to be a power of two.
    * Used as granularity for all page-related operations except for allocation,
    * which use the allocation_granularity(), see below.
    *
    * \returns the page size in bytes.
    */
@@ -568,20 +623,28 @@ class VirtualAddressSpace {
    */
   Address base() const { return base_; }
 
   /**
    * The size of the address space managed by this instance.
    *
    * \returns the size of this address space in bytes.
    */
   size_t size() const { return size_; }
 
+  /**
+   * The maximum page permissions that pages allocated inside this space can
+   * obtain.
+   *
+   * \returns the maximum page permissions.
+   */
+  PagePermissions max_page_permissions() const { return max_page_permissions_; }
+
   /**
    * Sets the random seed so that GetRandomPageAddress() will generate
    * repeatable sequences of random addresses.
    *
    * \param The seed for the PRNG.
    */
   virtual void SetRandomSeed(int64_t seed) = 0;
 
   /**
    * Returns a random address inside this address space, suitable for page
@@ -673,20 +736,57 @@ class VirtualAddressSpace {
    * invocation of AllocateGuardRegion.
    *
    * \param size The size in bytes of the guard region to free. This must match
    * the size passed to AllocateGuardRegion when the region was created.
    *
    * \returns true on success, false otherwise.
    */
   virtual V8_WARN_UNUSED_RESULT bool FreeGuardRegion(Address address,
                                                      size_t size) = 0;
 
+  /**
+   * Allocates shared memory pages with the given permissions.
+   *
+   * \param hint Placement hint. See AllocatePages.
+   *
+   * \param size The size of the allocation in bytes. Must be a multiple of the
+   * allocation_granularity().
+   *
+   * \param permissions The page permissions of the newly allocated pages.
+   *
+   * \param handle A platform-specific handle to a shared memory object. See
+   * the SharedMemoryHandleFromX routines above for ways to obtain these.
+   *
+   * \param offset The offset in the shared memory object at which the mapping
+   * should start. Must be a multiple of the allocation_granularity().
+   *
+   * \returns the start address of the allocated pages on success, zero on
+   * failure.
+   */
+  virtual V8_WARN_UNUSED_RESULT Address
+  AllocateSharedPages(Address hint, size_t size, PagePermissions permissions,
+                      PlatformSharedMemoryHandle handle, uint64_t offset) = 0;
+
+  /**
+   * Frees previously allocated shared pages.
+   *
+   * \param address The start address of the pages to free. This address must
+   * have been obtains from a call to AllocateSharedPages.
+   *
+   * \param size The size in bytes of the region to free. This must match the
+   * size passed to AllocateSharedPages when the pages were allocated.
+   *
+   * \returns true on success, false otherwise.
+   */
+  virtual V8_WARN_UNUSED_RESULT bool FreeSharedPages(Address address,
+                                                     size_t size) = 0;
+
   /**
    * Whether this instance can allocate subspaces or not.
    *
    * \returns true if subspaces can be allocated, false if not.
    */
   virtual bool CanAllocateSubspaces() = 0;
 
   /*
    * Allocate a subspace.
    *
@@ -696,28 +796,28 @@ class VirtualAddressSpace {
    *
    * \param hint Hints where the subspace should be allocated. See
    * AllocatePages() for more details.
    *
    * \param size The size in bytes of the subspace. Must be a multiple of the
    * allocation_granularity().
    *
    * \param alignment The alignment of the subspace in bytes. Must be a multiple
    * of the allocation_granularity() and should be a power of two.
    *
-   * \param max_permissions The maximum permissions that pages allocated in the
-   * subspace can obtain.
+   * \param max_page_permissions The maximum permissions that pages allocated in
+   * the subspace can obtain.
    *
    * \returns a new subspace or nullptr on failure.
    */
   virtual std::unique_ptr<VirtualAddressSpace> AllocateSubspace(
       Address hint, size_t size, size_t alignment,
-      PagePermissions max_permissions) = 0;
+      PagePermissions max_page_permissions) = 0;
 
   //
   // TODO(v8) maybe refactor the methods below before stabilizing the API. For
   // example by combining them into some form of page operation method that
   // takes a command enum as parameter.
   //
 
   /**
    * Frees memory in the given [address, address + size) range. address and
    * size should be aligned to the page_size(). The next write to this memory
@@ -743,20 +843,21 @@ class VirtualAddressSpace {
    * \returns true on success, false otherwise.
    */
   virtual V8_WARN_UNUSED_RESULT bool DecommitPages(Address address,
                                                    size_t size) = 0;
 
  private:
   const size_t page_size_;
   const size_t allocation_granularity_;
   const Address base_;
   const size_t size_;
+  const PagePermissions max_page_permissions_;
 };
 
 /**
  * V8 Allocator used for allocating zone backings.
  */
 class ZoneBackingAllocator {
  public:
   using MallocFn = void* (*)(size_t);
   using FreeFn = void (*)(void*);
 
