diff --git a/include/v8-platform.h b/include/v8-platform.h
index e27d26cb692..fc9a357feb6 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -174,23 +174,22 @@ class JobDelegate {
   /**
    * Returns a task_id unique among threads currently running this job, such
    * that GetTaskId() < worker count. To achieve this, the same task_id may be
    * reused by a different thread after a worker_task returns.
    */
   virtual uint8_t GetTaskId() = 0;
 
   /**
    * Returns true if the current task is called from the thread currently
    * running JobHandle::Join().
-   * TODO(etiennep): Make pure virtual once custom embedders implement it.
    */
-  virtual bool IsJoiningThread() const { return false; }
+  virtual bool IsJoiningThread() const = 0;
 };
 
 /**
  * Handle returned when posting a Job. Provides methods to control execution of
  * the posted Job.
  */
 class JobHandle {
  public:
   virtual ~JobHandle() = default;
 
@@ -213,44 +212,35 @@ class JobHandle {
   /**
    * Forces all existing workers to yield ASAP. Waits until they have all
    * returned from the Job's callback before returning.
    */
   virtual void Cancel() = 0;
 
   /*
    * Forces all existing workers to yield ASAP but doesn’t wait for them.
    * Warning, this is dangerous if the Job's callback is bound to or has access
    * to state which may be deleted after this call.
-   * TODO(etiennep): Cleanup once implemented by all embedders.
    */
-  virtual void CancelAndDetach() { Cancel(); }
+  virtual void CancelAndDetach() = 0;
 
   /**
    * Returns true if there's any work pending or any worker running.
    */
   virtual bool IsActive() = 0;
 
-  // TODO(etiennep): Clean up once all overrides are removed.
-  V8_DEPRECATED("Use !IsActive() instead.")
-  virtual bool IsCompleted() { return !IsActive(); }
-
   /**
    * Returns true if associated with a Job and other methods may be called.
    * Returns false after Join() or Cancel() was called. This may return true
    * even if no workers are running and IsCompleted() returns true
    */
   virtual bool IsValid() = 0;
 
-  // TODO(etiennep): Clean up once all overrides are removed.
-  V8_DEPRECATED("Use IsValid() instead.")
-  virtual bool IsRunning() { return IsValid(); }
-
   /**
    * Returns true if job priority can be changed.
    */
   virtual bool UpdatePriorityEnabled() const { return false; }
 
   /**
    *  Update this Job's priority.
    */
   virtual void UpdatePriority(TaskPriority new_priority) {}
 };
@@ -265,24 +255,20 @@ class JobTask {
   virtual void Run(JobDelegate* delegate) = 0;
 
   /**
    * Controls the maximum number of threads calling Run() concurrently, given
    * the number of threads currently assigned to this job and executing Run().
    * Run() is only invoked if the number of threads previously running Run() was
    * less than the value returned. Since GetMaxConcurrency() is a leaf function,
    * it must not call back any JobHandle methods.
    */
   virtual size_t GetMaxConcurrency(size_t worker_count) const = 0;
-
-  // TODO(1114823): Clean up once all overrides are removed.
-  V8_DEPRECATED("Use the version that takes |worker_count|.")
-  virtual size_t GetMaxConcurrency() const { return 0; }
 };
 
 /**
  * The interface represents complex arguments to trace events.
  */
 class ConvertableToTraceFormat {
  public:
   virtual ~ConvertableToTraceFormat() = default;
 
   /**
@@ -401,21 +387,20 @@ class PageAllocator {
    */
   virtual void* GetRandomMmapAddr() = 0;
 
   /**
    * Memory permissions.
    */
   enum Permission {
     kNoAccess,
     kRead,
     kReadWrite,
-    // TODO(hpayer): Remove this flag. Memory should never be rwx.
     kReadWriteExecute,
     kReadExecute,
     // Set this when reserving memory that will later require kReadWriteExecute
     // permissions. The resulting behavior is platform-specific, currently
     // this is used to set the MAP_JIT flag on Apple Silicon.
     // TODO(jkummerow): Remove this when Wasm has a platform-independent
     // w^x implementation.
     kNoAccessWillJitLater
   };
 
