diff --git a/include/v8-platform.h b/include/v8-platform.h
index 5110f432077..a3a939729c9 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -7,21 +7,21 @@
 
 #include <math.h>
 #include <stddef.h>
 #include <stdint.h>
 #include <stdlib.h>  // For abort.
 
 #include <memory>
 #include <string>
 
 #include "v8-source-location.h"  // NOLINT(build/include_directory)
-#include "v8config.h"  // NOLINT(build/include_directory)
+#include "v8config.h"            // NOLINT(build/include_directory)
 
 namespace v8 {
 
 class Isolate;
 
 // Valid priorities supported by the task scheduling infrastructure.
 enum class TaskPriority : uint8_t {
   /**
    * Best effort tasks are not critical for performance of the application. The
    * platform implementation should preempt such tasks if higher priority tasks
@@ -72,23 +72,22 @@ class IdleTask {
  * sequence. Tasks can be posted from any thread.
  */
 class TaskRunner {
  public:
   /**
    * Schedules a task to be invoked by this TaskRunner. The TaskRunner
    * implementation takes ownership of |task|.
    *
    * Embedders should override PostTaskImpl instead of this.
    */
-  void PostTask(
-      std::unique_ptr<Task> task,
-      const SourceLocation& location = SourceLocation::Current()) {
+  void PostTask(std::unique_ptr<Task> task,
+                const SourceLocation& location = SourceLocation::Current()) {
     PostTaskImpl(std::move(task), location);
   }
 
   /**
    * Schedules a task to be invoked by this TaskRunner. The TaskRunner
    * implementation takes ownership of |task|. The |task| cannot be nested
    * within other task executions.
    *
    * Tasks which shouldn't be interleaved with JS execution must be posted with
    * |PostNonNestableTask| or |PostNonNestableDelayedTask|. This is because the
@@ -546,20 +545,33 @@ class PageAllocator {
   /**
    * Decommits any wired memory pages in the given range, allowing the OS to
    * reclaim them, and marks the region as inacessible (kNoAccess). The address
    * range stays reserved and can be accessed again later by changing its
    * permissions. However, in that case the memory content is guaranteed to be
    * zero-initialized again. The memory must have been previously allocated by a
    * call to AllocatePages. Returns true on success, false otherwise.
    */
   virtual bool DecommitPages(void* address, size_t size) = 0;
 
+  /**
+   * Block any modifications to the given mapping such as changing permissions
+   * or unmapping the pages on supported platforms.
+   * The address space reservation will exist until the process ends, but it's
+   * possible to release the memory using DiscardSystemPages. Note that this
+   * might require write permissions to the page as e.g. on Linux, mseal will
+   * block discarding sealed anonymous memory.
+   */
+  virtual bool SealPages(void* address, size_t length) {
+    // TODO(360048056): make it pure once it's implemented on Chromium side.
+    return false;
+  }
+
   /**
    * INTERNAL ONLY: This interface has not been stabilised and may change
    * without notice from one release to another without being deprecated first.
    */
   class SharedMemoryMapping {
    public:
     // Implementations are expected to free the shared memory mapping in the
     // destructor.
     virtual ~SharedMemoryMapping() = default;
     virtual void* GetMemory() const = 0;
@@ -1079,38 +1091,32 @@ class Platform {
    * estimate the number of tasks a work package should be split into. A return
    * value of 0 means that there are no worker threads available. Note that a
    * value of 0 won't prohibit V8 from posting tasks using |CallOnWorkerThread|.
    */
   virtual int NumberOfWorkerThreads() = 0;
 
   /**
    * Returns a TaskRunner which can be used to post a task on the foreground.
    * The TaskRunner's NonNestableTasksEnabled() must be true. This function
    * should only be called from a foreground thread.
-   * TODO(chromium:1448758): Deprecate once |GetForegroundTaskRunner(Isolate*,
-   * TaskPriority)| is ready.
    */
-  virtual std::shared_ptr<v8::TaskRunner> GetForegroundTaskRunner(
-      Isolate* isolate) {
+  std::shared_ptr<v8::TaskRunner> GetForegroundTaskRunner(Isolate* isolate) {
     return GetForegroundTaskRunner(isolate, TaskPriority::kUserBlocking);
   }
 
   /**
    * Returns a TaskRunner with a specific |priority| which can be used to post a
    * task on the foreground thread. The TaskRunner's NonNestableTasksEnabled()
    * must be true. This function should only be called from a foreground thread.
-   * TODO(chromium:1448758): Make pure virtual once embedders implement it.
    */
   virtual std::shared_ptr<v8::TaskRunner> GetForegroundTaskRunner(
-      Isolate* isolate, TaskPriority priority) {
-    return nullptr;
-  }
+      Isolate* isolate, TaskPriority priority) = 0;
 
   /**
    * Schedules a task to be invoked on a worker thread.
    * Embedders should override PostTaskOnWorkerThreadImpl() instead of
    * CallOnWorkerThread().
    */
   void CallOnWorkerThread(
       std::unique_ptr<Task> task,
       const SourceLocation& location = SourceLocation::Current()) {
     PostTaskOnWorkerThreadImpl(TaskPriority::kUserVisible, std::move(task),
