diff --git a/include/v8-platform.h b/include/v8-platform.h
index 03be1c8cdd6..aae381b0806 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -163,20 +163,28 @@ class JobDelegate {
    * YieldIfNeeded()) as often as is reasonable.
    */
   virtual bool ShouldYield() = 0;
 
   /**
    * Notifies the scheduler that max concurrency was increased, and the number
    * of worker should be adjusted accordingly. See Platform::PostJob() for more
    * details.
    */
   virtual void NotifyConcurrencyIncrease() = 0;
+
+  /**
+   * Returns a task_id unique among threads currently running this job, such
+   * that GetTaskId() < worker count. To achieve this, the same task_id may be
+   * reused by a different thread after a worker_task returns.
+   * TODO(etiennep): Make pure virtual once custom embedders implement it.
+   */
+  virtual uint8_t GetTaskId() { return 0; }
 };
 
 /**
  * Handle returned when posting a Job. Provides methods to control execution of
  * the posted Job.
  */
 class JobHandle {
  public:
   virtual ~JobHandle() = default;
 
@@ -195,20 +203,26 @@ class JobHandle {
    * priority.
    */
   virtual void Join() = 0;
 
   /**
    * Forces all existing workers to yield ASAP. Waits until they have all
    * returned from the Job's callback before returning.
    */
   virtual void Cancel() = 0;
 
+  /**
+   * Returns true if there's no work pending and no worker running.
+   * TODO(etiennep): Make pure virtual once custom embedders implement it.
+   */
+  virtual bool IsCompleted() { return true; }
+
   /**
    * Returns true if associated with a Job and other methods may be called.
    * Returns false after Join() or Cancel() was called.
    */
   virtual bool IsRunning() = 0;
 };
 
 /**
  * A JobTask represents work to run in parallel from Platform::PostJob().
  */
@@ -218,20 +232,31 @@ class JobTask {
 
   virtual void Run(JobDelegate* delegate) = 0;
 
   /**
    * Controls the maximum number of threads calling Run() concurrently. Run() is
    * only invoked if the number of threads previously running Run() was less
    * than the value returned. Since GetMaxConcurrency() is a leaf function, it
    * must not call back any JobHandle methods.
    */
   virtual size_t GetMaxConcurrency() const = 0;
+
+  /*
+   * Meant to replace the version above, given the number of threads currently
+   * assigned to this job and executing Run(). This is useful when the result
+   * must include local work items not visible globaly by other workers.
+   * TODO(etiennep): Replace the version above by this once custom embedders are
+   * migrated.
+   */
+  size_t GetMaxConcurrency(size_t worker_count) const {
+    return GetMaxConcurrency();
+  }
 };
 
 /**
  * The interface represents complex arguments to trace events.
  */
 class ConvertableToTraceFormat {
  public:
   virtual ~ConvertableToTraceFormat() = default;
 
   /**
@@ -384,20 +409,83 @@ class PageAllocator {
    */
   virtual bool SetPermissions(void* address, size_t length,
                               Permission permissions) = 0;
 
   /**
    * Frees memory in the given [address, address + size) range. address and size
    * should be operating system page-aligned. The next write to this
    * memory area brings the memory transparently back.
    */
   virtual bool DiscardSystemPages(void* address, size_t size) { return true; }
+
+  /**
+   * INTERNAL ONLY: This interface has not been stabilised and may change
+   * without notice from one release to another without being deprecated first.
+   */
+  class SharedMemoryMapping {
+   public:
+    // Implementations are expected to free the shared memory mapping in the
+    // destructor.
+    virtual ~SharedMemoryMapping() = default;
+    virtual void* GetMemory() const = 0;
+  };
+
+  /**
+   * INTERNAL ONLY: This interface has not been stabilised and may change
+   * without notice from one release to another without being deprecated first.
+   */
+  class SharedMemory {
+   public:
+    // Implementations are expected to free the shared memory in the destructor.
+    virtual ~SharedMemory() = default;
+    virtual std::unique_ptr<SharedMemoryMapping> RemapTo(
+        void* new_address) const = 0;
+    virtual void* GetMemory() const = 0;
+    virtual size_t GetSize() const = 0;
+  };
+
+  /**
+   * INTERNAL ONLY: This interface has not been stabilised and may change
+   * without notice from one release to another without being deprecated first.
+   *
+   * Reserve pages at a fixed address returning whether the reservation is
+   * possible. The reserved memory is detached from the PageAllocator and so
+   * should not be freed by it. It's intended for use with
+   * SharedMemory::RemapTo, where ~SharedMemoryMapping would free the memory.
+   */
+  virtual bool ReserveForSharedMemoryMapping(void* address, size_t size) {
+    return false;
+  }
+
+  /**
+   * INTERNAL ONLY: This interface has not been stabilised and may change
+   * without notice from one release to another without being deprecated first.
+   *
+   * Allocates shared memory pages. Not all PageAllocators need support this and
+   * so this method need not be overridden.
+   * Allocates a new read-only shared memory region of size |length| and copies
+   * the memory at |original_address| into it.
+   */
+  virtual std::unique_ptr<SharedMemory> AllocateSharedPages(
+      size_t length, const void* original_address) {
+    return {};
+  }
+
+  /**
+   * INTERNAL ONLY: This interface has not been stabilised and may change
+   * without notice from one release to another without being deprecated first.
+   *
+   * If not overridden and changed to return true, V8 will not attempt to call
+   * AllocateSharedPages or RemapSharedPages. If overridden, AllocateSharedPages
+   * and RemapSharedPages must also be overridden.
+   */
+  virtual bool CanAllocateSharedPages() { return false; }
 };
 
 /**
  * V8 Platform abstraction layer.
  *
  * The embedder has to provide an implementation of this interface before
  * initializing the rest of V8.
  */
 class Platform {
  public:
@@ -533,23 +621,21 @@ class Platform {
    *
    * A sufficient PostJob() implementation that uses the default Job provided in
    * libplatform looks like:
    *  std::unique_ptr<JobHandle> PostJob(
    *      TaskPriority priority, std::unique_ptr<JobTask> job_task) override {
    *    return v8::platform::NewDefaultJobHandle(
    *        this, priority, std::move(job_task), NumberOfWorkerThreads());
    * }
    */
   virtual std::unique_ptr<JobHandle> PostJob(
-      TaskPriority priority, std::unique_ptr<JobTask> job_task) {
-    return nullptr;
-  }
+      TaskPriority priority, std::unique_ptr<JobTask> job_task) = 0;
 
   /**
    * Monotonically increasing time in seconds from an arbitrary fixed point in
    * the past. This function is expected to return at least
    * millisecond-precision values. For this reason,
    * it is recommended that the fixed point be no further in the past than
    * the epoch.
    **/
   virtual double MonotonicallyIncreasingTime() = 0;
 
