diff --git a/include/v8-platform.h b/include/v8-platform.h
index b61f27af6e3..313c0287bcf 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -69,76 +69,98 @@ class IdleTask {
  * A TaskRunner allows scheduling of tasks. The TaskRunner may still be used to
  * post tasks after the isolate gets destructed, but these tasks may not get
  * executed anymore. All tasks posted to a given TaskRunner will be invoked in
  * sequence. Tasks can be posted from any thread.
  */
 class TaskRunner {
  public:
   /**
    * Schedules a task to be invoked by this TaskRunner. The TaskRunner
    * implementation takes ownership of |task|.
+   *
+   * Embedders should override PostTaskImpl instead of this.
    */
-  virtual void PostTask(std::unique_ptr<Task> task) = 0;
+  virtual void PostTask(std::unique_ptr<Task> task) {
+    PostTaskImpl(std::move(task), SourceLocation::Current());
+  }
 
   /**
    * Schedules a task to be invoked by this TaskRunner. The TaskRunner
    * implementation takes ownership of |task|. The |task| cannot be nested
    * within other task executions.
    *
    * Tasks which shouldn't be interleaved with JS execution must be posted with
    * |PostNonNestableTask| or |PostNonNestableDelayedTask|. This is because the
    * embedder may process tasks in a callback which is called during JS
    * execution.
    *
    * In particular, tasks which execute JS must be non-nestable, since JS
    * execution is not allowed to nest.
    *
    * Requires that |TaskRunner::NonNestableTasksEnabled()| is true.
+   *
+   * Embedders should override PostNonNestableTaskImpl instead of this.
    */
-  virtual void PostNonNestableTask(std::unique_ptr<Task> task) {}
+  virtual void PostNonNestableTask(std::unique_ptr<Task> task) {
+    PostNonNestableTaskImpl(std::move(task), SourceLocation::Current());
+  }
 
   /**
    * Schedules a task to be invoked by this TaskRunner. The task is scheduled
    * after the given number of seconds |delay_in_seconds|. The TaskRunner
    * implementation takes ownership of |task|.
+   *
+   * Embedders should override PostDelayedTaskImpl instead of this.
    */
   virtual void PostDelayedTask(std::unique_ptr<Task> task,
-                               double delay_in_seconds) = 0;
+                               double delay_in_seconds) {
+    PostDelayedTaskImpl(std::move(task), delay_in_seconds,
+                        SourceLocation::Current());
+  }
 
   /**
    * Schedules a task to be invoked by this TaskRunner. The task is scheduled
    * after the given number of seconds |delay_in_seconds|. The TaskRunner
    * implementation takes ownership of |task|. The |task| cannot be nested
    * within other task executions.
    *
    * Tasks which shouldn't be interleaved with JS execution must be posted with
    * |PostNonNestableTask| or |PostNonNestableDelayedTask|. This is because the
    * embedder may process tasks in a callback which is called during JS
    * execution.
    *
    * In particular, tasks which execute JS must be non-nestable, since JS
    * execution is not allowed to nest.
    *
    * Requires that |TaskRunner::NonNestableDelayedTasksEnabled()| is true.
+   *
+   * Embedders should override PostNonNestableDelayedTaskImpl instead of this.
    */
   virtual void PostNonNestableDelayedTask(std::unique_ptr<Task> task,
-                                          double delay_in_seconds) {}
+                                          double delay_in_seconds) {
+    PostNonNestableDelayedTaskImpl(std::move(task), delay_in_seconds,
+                                   SourceLocation::Current());
+  }
 
   /**
    * Schedules an idle task to be invoked by this TaskRunner. The task is
    * scheduled when the embedder is idle. Requires that
    * |TaskRunner::IdleTasksEnabled()| is true. Idle tasks may be reordered
    * relative to other task types and may be starved for an arbitrarily long
    * time if no idle time is available. The TaskRunner implementation takes
    * ownership of |task|.
+   *
+   * Embedders should override PostIdleTaskImpl instead of this.
    */
-  virtual void PostIdleTask(std::unique_ptr<IdleTask> task) = 0;
+  virtual void PostIdleTask(std::unique_ptr<IdleTask> task) {
+    PostIdleTaskImpl(std::move(task), SourceLocation::Current());
+  }
 
   /**
    * Returns true if idle tasks are enabled for this TaskRunner.
    */
   virtual bool IdleTasksEnabled() = 0;
 
   /**
    * Returns true if non-nestable tasks are enabled for this TaskRunner.
    */
   virtual bool NonNestableTasksEnabled() const { return false; }
@@ -146,20 +168,37 @@ class TaskRunner {
   /**
    * Returns true if non-nestable delayed tasks are enabled for this TaskRunner.
    */
   virtual bool NonNestableDelayedTasksEnabled() const { return false; }
 
   TaskRunner() = default;
   virtual ~TaskRunner() = default;
 
   TaskRunner(const TaskRunner&) = delete;
   TaskRunner& operator=(const TaskRunner&) = delete;
+
+ protected:
+  /**
+   * Implementation of above methods with an additional `location` argument.
+   */
+  virtual void PostTaskImpl(std::unique_ptr<Task> task,
+                            const SourceLocation& location) {}
+  virtual void PostNonNestableTaskImpl(std::unique_ptr<Task> task,
+                                       const SourceLocation& location) {}
+  virtual void PostDelayedTaskImpl(std::unique_ptr<Task> task,
+                                   double delay_in_seconds,
+                                   const SourceLocation& location) {}
+  virtual void PostNonNestableDelayedTaskImpl(std::unique_ptr<Task> task,
+                                              double delay_in_seconds,
+                                              const SourceLocation& location) {}
+  virtual void PostIdleTaskImpl(std::unique_ptr<IdleTask> task,
+                                const SourceLocation& location) {}
 };
 
 /**
  * Delegate that's passed to Job's worker task, providing an entry point to
  * communicate with the scheduler.
  */
 class JobDelegate {
  public:
   /**
    * Returns true if this thread *must* return from the worker task on the
