diff --git a/include/v8-platform.h b/include/v8-platform.h
index 2decd2088b5..91b3fd9cc3f 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -515,40 +515,40 @@ class PageAllocator {
 // Opaque type representing a handle to a shared memory region.
 using PlatformSharedMemoryHandle = intptr_t;
 static constexpr PlatformSharedMemoryHandle kInvalidSharedMemoryHandle = -1;
 
 // Conversion routines from the platform-dependent shared memory identifiers
 // into the opaque PlatformSharedMemoryHandle type. These use the underlying
 // types (e.g. unsigned int) instead of the typedef'd ones (e.g. mach_port_t)
 // to avoid pulling in large OS header files into this header file. Instead,
 // the users of these routines are expected to include the respecitve OS
 // headers in addition to this one.
-#if defined(V8_OS_MACOSX) && !defined(V8_OS_IOS)
+#if V8_OS_MACOS
 // Convert between a shared memory handle and a mach_port_t referencing a memory
 // entry object.
 inline PlatformSharedMemoryHandle SharedMemoryHandleFromMachMemoryEntry(
     unsigned int port) {
   return static_cast<PlatformSharedMemoryHandle>(port);
 }
 inline unsigned int MachMemoryEntryFromSharedMemoryHandle(
     PlatformSharedMemoryHandle handle) {
   return static_cast<unsigned int>(handle);
 }
-#elif defined(V8_OS_FUCHSIA)
+#elif V8_OS_FUCHSIA
 // Convert between a shared memory handle and a zx_handle_t to a VMO.
 inline PlatformSharedMemoryHandle SharedMemoryHandleFromVMO(uint32_t handle) {
   return static_cast<PlatformSharedMemoryHandle>(handle);
 }
 inline uint32_t VMOFromSharedMemoryHandle(PlatformSharedMemoryHandle handle) {
   return static_cast<uint32_t>(handle);
 }
-#elif defined(V8_OS_WIN)
+#elif V8_OS_WIN
 // Convert between a shared memory handle and a Windows HANDLE to a file mapping
 // object.
 inline PlatformSharedMemoryHandle SharedMemoryHandleFromFileMapping(
     void* handle) {
   return reinterpret_cast<PlatformSharedMemoryHandle>(handle);
 }
 inline void* FileMappingFromSharedMemoryHandle(
     PlatformSharedMemoryHandle handle) {
   return reinterpret_cast<void*>(handle);
 }
@@ -574,20 +574,22 @@ enum class PagePermissions {
   kReadExecute,
 };
 
 /**
  * Class to manage a virtual memory address space.
  *
  * This class represents a contiguous region of virtual address space in which
  * sub-spaces and (private or shared) memory pages can be allocated, freed, and
  * modified. This interface is meant to eventually replace the PageAllocator
  * interface, and can be used as an alternative in the meantime.
+ *
+ * This API is not yet stable and may change without notice!
  */
 class VirtualAddressSpace {
  public:
   using Address = uintptr_t;
 
   VirtualAddressSpace(size_t page_size, size_t allocation_granularity,
                       Address base, size_t size,
                       PagePermissions max_page_permissions)
       : page_size_(page_size),
         allocation_granularity_(allocation_granularity),
@@ -675,30 +677,30 @@ class VirtualAddressSpace {
    * failure.
    */
   static constexpr Address kNoHint = 0;
   virtual V8_WARN_UNUSED_RESULT Address
   AllocatePages(Address hint, size_t size, size_t alignment,
                 PagePermissions permissions) = 0;
 
   /**
    * Frees previously allocated pages.
    *
+   * This function will terminate the process on failure as this implies a bug
+   * in the client. As such, there is no return value.
+   *
    * \param address The start address of the pages to free. This address must
-   * have been obtains from a call to AllocatePages.
+   * have been obtained through a call to AllocatePages.
    *
    * \param size The size in bytes of the region to free. This must match the
    * size passed to AllocatePages when the pages were allocated.
-   *
-   * \returns true on success, false otherwise.
    */
-  virtual V8_WARN_UNUSED_RESULT bool FreePages(Address address,
-                                               size_t size) = 0;
+  virtual void FreePages(Address address, size_t size) = 0;
 
   /**
    * Sets permissions of all allocated pages in the given range.
    *
    * \param address The start address of the range. Must be aligned to
    * page_size().
    *
    * \param size The size in bytes of the range. Must be a multiple
    * of page_size().
    *
@@ -724,31 +726,31 @@ class VirtualAddressSpace {
    * the allocation_granularity().
    *
    * \returns true on success, false otherwise.
    */
   virtual V8_WARN_UNUSED_RESULT bool AllocateGuardRegion(Address address,
                                                          size_t size) = 0;
 
   /**
    * Frees an existing guard region.
    *
+   * This function will terminate the process on failure as this implies a bug
+   * in the client. As such, there is no return value.
+   *
    * \param address The start address of the guard region to free. This address
    * must have previously been used as address parameter in a successful
    * invocation of AllocateGuardRegion.
    *
    * \param size The size in bytes of the guard region to free. This must match
    * the size passed to AllocateGuardRegion when the region was created.
-   *
-   * \returns true on success, false otherwise.
    */
-  virtual V8_WARN_UNUSED_RESULT bool FreeGuardRegion(Address address,
-                                                     size_t size) = 0;
+  virtual void FreeGuardRegion(Address address, size_t size) = 0;
 
   /**
    * Allocates shared memory pages with the given permissions.
    *
    * \param hint Placement hint. See AllocatePages.
    *
    * \param size The size of the allocation in bytes. Must be a multiple of the
    * allocation_granularity().
    *
    * \param permissions The page permissions of the newly allocated pages.
@@ -762,30 +764,30 @@ class VirtualAddressSpace {
    * \returns the start address of the allocated pages on success, zero on
    * failure.
    */
   virtual V8_WARN_UNUSED_RESULT Address
   AllocateSharedPages(Address hint, size_t size, PagePermissions permissions,
                       PlatformSharedMemoryHandle handle, uint64_t offset) = 0;
 
   /**
    * Frees previously allocated shared pages.
    *
+   * This function will terminate the process on failure as this implies a bug
+   * in the client. As such, there is no return value.
+   *
    * \param address The start address of the pages to free. This address must
-   * have been obtains from a call to AllocateSharedPages.
+   * have been obtained through a call to AllocateSharedPages.
    *
    * \param size The size in bytes of the region to free. This must match the
    * size passed to AllocateSharedPages when the pages were allocated.
-   *
-   * \returns true on success, false otherwise.
    */
-  virtual V8_WARN_UNUSED_RESULT bool FreeSharedPages(Address address,
-                                                     size_t size) = 0;
+  virtual void FreeSharedPages(Address address, size_t size) = 0;
 
   /**
    * Whether this instance can allocate subspaces or not.
    *
    * \returns true if subspaces can be allocated, false if not.
    */
   virtual bool CanAllocateSubspaces() = 0;
 
   /*
    * Allocate a subspace.
