diff --git a/include/v8-platform.h b/include/v8-platform.h
index 1f1497f6ccd..e27d26cb692 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -218,36 +218,38 @@ class JobHandle {
 
   /*
    * Forces all existing workers to yield ASAP but doesn’t wait for them.
    * Warning, this is dangerous if the Job's callback is bound to or has access
    * to state which may be deleted after this call.
    * TODO(etiennep): Cleanup once implemented by all embedders.
    */
   virtual void CancelAndDetach() { Cancel(); }
 
   /**
-   * Returns true if there's currently no work pending and no worker running.
-   * TODO(etiennep): Deprecate IsCompleted in favor of IsActive once implemented
-   * by all embedders.
+   * Returns true if there's any work pending or any worker running.
    */
-  virtual bool IsCompleted() = 0;
-  virtual bool IsActive() { return !IsCompleted(); }
+  virtual bool IsActive() = 0;
+
+  // TODO(etiennep): Clean up once all overrides are removed.
+  V8_DEPRECATED("Use !IsActive() instead.")
+  virtual bool IsCompleted() { return !IsActive(); }
 
   /**
    * Returns true if associated with a Job and other methods may be called.
    * Returns false after Join() or Cancel() was called. This may return true
    * even if no workers are running and IsCompleted() returns true
-   * TODO(etiennep): Deprecate IsRunning in favor of IsValid once implemented by
-   * all embedders.
    */
-  virtual bool IsRunning() = 0;
-  virtual bool IsValid() { return IsRunning(); }
+  virtual bool IsValid() = 0;
+
+  // TODO(etiennep): Clean up once all overrides are removed.
+  V8_DEPRECATED("Use IsValid() instead.")
+  virtual bool IsRunning() { return IsValid(); }
 
   /**
    * Returns true if job priority can be changed.
    */
   virtual bool UpdatePriorityEnabled() const { return false; }
 
   /**
    *  Update this Job's priority.
    */
   virtual void UpdatePriority(TaskPriority new_priority) {}
