diff --git a/include/v8-platform.h b/include/v8-platform.h
index 534d73e3f7b..5d23cd665ed 100644
--- a/include/v8-platform.h
+++ b/include/v8-platform.h
@@ -319,21 +319,22 @@ class Platform {
    * Gets the number of worker threads used by
    * Call(BlockingTask)OnWorkerThread(). This can be used to estimate the number
    * of tasks a work package should be split into. A return value of 0 means
    * that there are no worker threads available. Note that a value of 0 won't
    * prohibit V8 from posting tasks using |CallOnWorkerThread|.
    */
   virtual int NumberOfWorkerThreads() = 0;
 
   /**
    * Returns a TaskRunner which can be used to post a task on the foreground.
-   * This function should only be called from a foreground thread.
+   * The TaskRunner's NonNestableTasksEnabled() must be true. This function
+   * should only be called from a foreground thread.
    */
   virtual std::shared_ptr<v8::TaskRunner> GetForegroundTaskRunner(
       Isolate* isolate) = 0;
 
   /**
    * Schedules a task to be invoked on a worker thread.
    */
   virtual void CallOnWorkerThread(std::unique_ptr<Task> task) = 0;
 
   /**
@@ -355,53 +356,20 @@ class Platform {
     CallOnWorkerThread(std::move(task));
   }
 
   /**
    * Schedules a task to be invoked on a worker thread after |delay_in_seconds|
    * expires.
    */
   virtual void CallDelayedOnWorkerThread(std::unique_ptr<Task> task,
                                          double delay_in_seconds) = 0;
 
-  /**
-   * Schedules a task to be invoked on a foreground thread wrt a specific
-   * |isolate|. Tasks posted for the same isolate should be execute in order of
-   * scheduling. The definition of "foreground" is opaque to V8.
-   */
-  V8_DEPRECATED("Use a taskrunner acquired by GetForegroundTaskRunner instead.")
-  virtual void CallOnForegroundThread(Isolate* isolate, Task* task) { abort(); }
-
-  /**
-   * Schedules a task to be invoked on a foreground thread wrt a specific
-   * |isolate| after the given number of seconds |delay_in_seconds|.
-   * Tasks posted for the same isolate should be execute in order of
-   * scheduling. The definition of "foreground" is opaque to V8.
-   */
-  V8_DEPRECATED("Use a taskrunner acquired by GetForegroundTaskRunner instead.")
-  virtual void CallDelayedOnForegroundThread(Isolate* isolate, Task* task,
-                                             double delay_in_seconds) {
-    abort();
-  }
-
-  /**
-   * Schedules a task to be invoked on a foreground thread wrt a specific
-   * |isolate| when the embedder is idle.
-   * Requires that SupportsIdleTasks(isolate) is true.
-   * Idle tasks may be reordered relative to other task types and may be
-   * starved for an arbitrarily long time if no idle time is available.
-   * The definition of "foreground" is opaque to V8.
-   */
-  V8_DEPRECATED("Use a taskrunner acquired by GetForegroundTaskRunner instead.")
-  virtual void CallIdleOnForegroundThread(Isolate* isolate, IdleTask* task) {
-    abort();
-  }
-
   /**
    * Returns true if idle tasks are enabled for the given |isolate|.
    */
   virtual bool IdleTasksEnabled(Isolate* isolate) { return false; }
 
   /**
    * Monotonically increasing time in seconds from an arbitrary fixed point in
    * the past. This function is expected to return at least
    * millisecond-precision values. For this reason,
    * it is recommended that the fixed point be no further in the past than
