diff --git a/include/v8-template.h b/include/v8-template.h
index 4d8386865e6..a7836120737 100644
--- a/include/v8-template.h
+++ b/include/v8-template.h
@@ -53,33 +53,55 @@ class V8_EXPORT Template : public Data {
    *
    * The property must be defined either as a primitive value, or a template.
    */
   void Set(Local<Name> name, Local<Data> value,
            PropertyAttribute attributes = None);
   void SetPrivate(Local<Private> name, Local<Data> value,
                   PropertyAttribute attributes = None);
   V8_INLINE void Set(Isolate* isolate, const char* name, Local<Data> value,
                      PropertyAttribute attributes = None);
 
+  /**
+   * Sets an "accessor property" on the object template, see
+   * https://tc39.es/ecma262/#sec-object-type.
+   *
+   * Whenever the property with the given name is accessed on objects
+   * created from this ObjectTemplate the getter and setter functions
+   * are called.
+   *
+   * \param name The name of the property for which an accessor is added.
+   * \param getter The callback to invoke when getting the property.
+   * \param setter The callback to invoke when setting the property.
+   * \param data A piece of data that will be passed to the getter and setter
+   *   callbacks whenever they are invoked.
+   * \param attribute The attributes of the property for which an accessor
+   *   is added.
+   */
   void SetAccessorProperty(
       Local<Name> name,
       Local<FunctionTemplate> getter = Local<FunctionTemplate>(),
       Local<FunctionTemplate> setter = Local<FunctionTemplate>(),
       PropertyAttribute attribute = None);
 
   /**
+   * Sets a "data property" on the object template, see
+   * https://tc39.es/ecma262/#sec-object-type.
+   *
    * Whenever the property with the given name is accessed on objects
    * created from this Template the getter and setter callbacks
    * are called instead of getting and setting the property directly
    * on the JavaScript object.
+   * Note that in case a property is written via a "child" object, the setter
+   * will not be called according to the JavaScript specification. See
+   * https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-set-p-v-receiver.
    *
-   * \param name The name of the property for which an accessor is added.
+   * \param name The name of the data property for which an accessor is added.
    * \param getter The callback to invoke when getting the property.
    * \param setter The callback to invoke when setting the property.
    * \param data A piece of data that will be passed to the getter and setter
    *   callbacks whenever they are invoked.
    * \param attribute The attributes of the property for which an accessor
    *   is added.
    */
   void SetNativeDataProperty(
       Local<Name> name, AccessorNameGetterCallback getter,
       AccessorNameSetterCallback setter = nullptr,
@@ -885,21 +907,21 @@ struct IndexedPropertyHandlerConfiguration {
       : getter(reinterpret_cast<void*>(getter)),
         setter(reinterpret_cast<void*>(setter)),
         query(reinterpret_cast<void*>(query)),
         deleter(reinterpret_cast<void*>(deleter)),
         enumerator(enumerator),
         definer(reinterpret_cast<void*>(definer)),
         descriptor(reinterpret_cast<void*>(descriptor)),
         data(data),
         flags(WithNewSignatureFlag(flags)) {}
 
-  V8_DEPRECATE_SOON(
+  V8_DEPRECATED(
       "Provide interceptor callbacks with new signatures instead "
       "(IndexedPropertyXxxCallbackV2)")
   IndexedPropertyHandlerConfiguration(
       IndexedPropertyGetterCallback getter,          //
       IndexedPropertySetterCallback setter,          //
       IndexedPropertyQueryCallback query,            //
       IndexedPropertyDeleterCallback deleter,        //
       IndexedPropertyEnumeratorCallback enumerator,  //
       IndexedPropertyDefinerCallback definer,        //
       IndexedPropertyDescriptorCallback descriptor,  //
@@ -926,21 +948,21 @@ struct IndexedPropertyHandlerConfiguration {
       : getter(reinterpret_cast<void*>(getter)),
         setter(reinterpret_cast<void*>(setter)),
         query(reinterpret_cast<void*>(query)),
         deleter(reinterpret_cast<void*>(deleter)),
         enumerator(enumerator),
         definer(nullptr),
         descriptor(nullptr),
         data(data),
         flags(WithNewSignatureFlag(flags)) {}
 
-  V8_DEPRECATE_SOON(
+  V8_DEPRECATED(
       "Provide interceptor callbacks with new signatures instead "
       "(IndexedPropertyXxxCallbackV2)")
   explicit IndexedPropertyHandlerConfiguration(
       IndexedPropertyGetterCallback getter,
       IndexedPropertySetterCallback setter = nullptr,
       IndexedPropertyQueryCallback query = nullptr,
       IndexedPropertyDeleterCallback deleter = nullptr,
       IndexedPropertyEnumeratorCallback enumerator = nullptr,
       Local<Value> data = Local<Value>(),
       PropertyHandlerFlags flags = PropertyHandlerFlags::kNone)
@@ -966,21 +988,21 @@ struct IndexedPropertyHandlerConfiguration {
       : getter(reinterpret_cast<void*>(getter)),
         setter(reinterpret_cast<void*>(setter)),
         query(nullptr),
         deleter(reinterpret_cast<void*>(deleter)),
         enumerator(enumerator),
         definer(reinterpret_cast<void*>(definer)),
         descriptor(reinterpret_cast<void*>(descriptor)),
         data(data),
         flags(WithNewSignatureFlag(flags)) {}
 
-  V8_DEPRECATE_SOON(
+  V8_DEPRECATED(
       "Provide interceptor callbacks with new signatures instead "
       "(IndexedPropertyXxxCallbackV2)")
   IndexedPropertyHandlerConfiguration(
       IndexedPropertyGetterCallback getter,
       IndexedPropertySetterCallback setter,
       IndexedPropertyDescriptorCallback descriptor,
       IndexedPropertyDeleterCallback deleter,
       IndexedPropertyEnumeratorCallback enumerator,
       IndexedPropertyDefinerCallback definer,
       Local<Value> data = Local<Value>(),
@@ -1035,69 +1057,44 @@ class V8_EXPORT ObjectTemplate : public Template {
    * on the JavaScript object.
    *
    * \param name The name of the property for which an accessor is added.
    * \param getter The callback to invoke when getting the property.
    * \param setter The callback to invoke when setting the property.
    * \param data A piece of data that will be passed to the getter and setter
    *   callbacks whenever they are invoked.
    * \param attribute The attributes of the property for which an accessor
    *   is added.
    */
-  V8_DEPRECATE_SOON("Use SetNativeDataProperty instead")
+  V8_DEPRECATED(
+      "Use SetNativeDataProperty or SetAccessorProperty instead depending on "
+      "the required semantics. See http://crbug.com/336325111. This method "
+      "will be removed in V8 12.8.")
   void SetAccessor(
       Local<Name> name, AccessorNameGetterCallback getter,
       AccessorNameSetterCallback setter = nullptr,
       Local<Value> data = Local<Value>(), PropertyAttribute attribute = None,
       SideEffectType getter_side_effect_type = SideEffectType::kHasSideEffect,
       SideEffectType setter_side_effect_type = SideEffectType::kHasSideEffect);
 
   /**
    * Sets a named property handler on the object template.
    *
    * Whenever a property whose name is a string or a symbol is accessed on
    * objects created from this object template, the provided callback is
    * invoked instead of accessing the property directly on the JavaScript
    * object.
    *
    * @param configuration The NamedPropertyHandlerConfiguration that defines the
    * callbacks to invoke when accessing a property.
    */
   void SetHandler(const NamedPropertyHandlerConfiguration& configuration);
 
-  /**
-   * Sets an indexed property handler on the object template.
-   *
-   * Whenever an indexed property is accessed on objects created from
-   * this object template, the provided callback is invoked instead of
-   * accessing the property directly on the JavaScript object.
-   *
-   * \param getter The callback to invoke when getting a property.
-   * \param setter The callback to invoke when setting a property.
-   * \param query The callback to invoke to check if an object has a property.
-   * \param deleter The callback to invoke when deleting a property.
-   * \param enumerator The callback to invoke to enumerate all the indexed
-   *   properties of an object.
-   * \param data A piece of data that will be passed to the callbacks
-   *   whenever they are invoked.
-   */
-  V8_DEPRECATED("Use SetHandler instead")
-  void SetIndexedPropertyHandler(
-      IndexedPropertyGetterCallback getter,
-      IndexedPropertySetterCallback setter = nullptr,
-      IndexedPropertyQueryCallback query = nullptr,
-      IndexedPropertyDeleterCallback deleter = nullptr,
-      IndexedPropertyEnumeratorCallback enumerator = nullptr,
-      Local<Value> data = Local<Value>()) {
-    SetHandler(IndexedPropertyHandlerConfiguration(getter, setter, query,
-                                                   deleter, enumerator, data));
-  }
-
   /**
    * Sets an indexed property handler on the object template.
    *
    * Whenever an indexed property is accessed on objects created from
    * this object template, the provided callback is invoked instead of
    * accessing the property directly on the JavaScript object.
    *
    * @param configuration The IndexedPropertyHandlerConfiguration that defines
    * the callbacks to invoke when accessing a property.
    */
